staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go

```diff
@@ -0,0 +1,165 @@
+	}
+	// Close channel when detecting close connection.
+	closeHandler := conn.CloseHandler()
+	conn.SetCloseHandler(func(code int, text string) error {
+		klog.V(3).Infof("websocket conn close: %d--%s", code, text)
+		close(closeChan)
+		err := closeHandler(code, text)
+		return err
+	})
+	return tConn
+}
+
+func (c *TunnelingConnection) Read(p []byte) (int, error) {
+	klog.Infoln("tunneling connection read...")
+	defer klog.Infoln("tunneling connection read...complete")
+	for {
+		if c.inProgressMessage == nil {
+			klog.Infoln("tunneling connection read before NextReader()...")
+			messageType, nextReader, err := c.conn.NextReader()
+			if err != nil {
+				closeError := &gwebsocket.CloseError{}
+				if errors.As(err, &closeError) && closeError.Code == gwebsocket.CloseNormalClosure {
+					return 0, io.EOF
+				}
+				klog.Errorf("tunneling connection NextReader() error: %v", err)
+				return 0, err
+			}
+			if messageType != gwebsocket.BinaryMessage {
+				return 0, fmt.Errorf("invalid message type received")
+			}
+			c.inProgressMessage = nextReader
+		}
+
+		klog.Infoln("tunneling connection read in progress message...")
+		i, err := c.inProgressMessage.Read(p)
+		klog.Infof("tunneling connection read in progress message...%d bytes, (error: %v)", i, err)
+		switch {
+		case err == nil:
+			return i, nil
+		case err == io.EOF:
+			c.inProgressMessage = nil
+		case err != nil:
+			return i, err
+		}
+	}
+}
+
+func (c *TunnelingConnection) Write(p []byte) (int, error) {
+	klog.Infof("tunneling connection write: %d bytes", len(p))
+	defer klog.Infoln("tunneling connection write...complete")
+	if c.conn == nil {
+		return 0, fmt.Errorf("write on closed tunneling connection")
+	}
+	// err := c.SetWriteDeadline(time.Now().Add(writeDeadline))
+	// if err != nil {
+	// 	klog.V(4).Infof("Tunneling conn setting write deadline failed %v", err)
+	// 	return 0, err
+	// }
+	w, err := c.conn.NextWriter(gwebsocket.BinaryMessage)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		if w != nil {
+			w.Close() //nolint:errcheck
+		}
+	}()
+	// Throttle rate of writing to websocket connection (there is only one Writer
+	// that can be active on the connection at a time).
+	// if s.maxBytesPerSec > 0 {
+	// 	w = flowrate.NewWriter(w, s.maxBytesPerSec)
+	// }
+	// Next, write the passed data in "p".
+	n, err := w.Write(p)
+	if err != nil {
+		return n, err
+	}
+	err = w.Close()
+	w = nil
+	return n, err
+}
+
+func (c *TunnelingConnection) Close() error {
+	klog.Infoln("tunneling connection Close()...")
+	// Signal other endpoint that websocket connection is closing.
+	c.conn.WriteControl(gwebsocket.CloseMessage, []byte{}, time.Now().Add(writeDeadline)) //nolint:errcheck
+	return c.conn.Close()
+}
+
+func (c *TunnelingConnection) LocalAddr() net.Addr {
+	return nil
+}
+
+func (c *TunnelingConnection) RemoteAddr() net.Addr {
+	return nil
+}
+
+func (c *TunnelingConnection) SetDeadline(t time.Time) error {
+	err := c.SetReadDeadline(t)
```

Comment:
probably call both even if one errors, and return an error if either errored

does this do the right thing with the zero value?

> A zero value for t means I/O operations will not time out.

Reply:
Fixed.